### PR TITLE
add block-sparse fp8 prefill attention (dim=128, two quant schemes)

### PR DESCRIPTION
**Pull request by bcacdwk**
## Summary
  - Block-sparse paged-KV prefill attention for FP8, dim=128 (with kvcache)
  - Two quantization schemes via `quant_type` parameter:
    - `QPERTOKEN_PERHEAD_KPERTENSOR_VPERTENSOR` (default)
    - `QPERTOKEN_PERHEAD_KPERTOKEN_PERHEAD_VPERHEAD`
  - Updated existing dense FP8 prefill kernels to support both quant schemes.

## Test result
  - [x] `make wheel` builds clean on H20 / CUDA 13.0 / torch 2.10
  - [x] All 9 `test_attention_*.py` pass
  - [x] Full `pytest tests/` regression: all passed

## Benchmark
fig_bsa_sparsity_line_new

### CODE REVIEW

File: hpc/attention.py [hunk 1]
```diff
@@ -204,6 +245,153 @@ def attention_with_kvcache_prefill_fp8(
...
+        block_ids,
+        seqlens_kvcache,
+        max_seqlens_q,
+        quant_type.value,
+        block_mask,
+        output,
+    )
+
+
+def attention_blocksparse_prefill_fp8_dim192(
```
Comment (shaochangxu) [inline hpc/attention.py]: 这个entry dim192在async里dispatch开吧，调用的时候不用让用户感知dim

File: src/attention/prefill/config.h [hunk 2]
```diff
@@ -181,11 +181,194 @@ struct AttentionKVCachePrefillConfig {
...
+  using SLayoutK = decltype(tile_to_shape(SLayoutKAtom{},
+                                          make_shape(Int<kTileN>{}, Int<kTileK>{}, Int<kStage>{})));
+  using ShapeV =
+      Shape<Shape<_64, _2>, decltype(shape<0>(SLayoutK{})), decltype(shape<2>(SLayoutK{}))>;
+  using StrideV = decltype(make_identity_layout(SLayoutK{}.shape()).stride());
+  using TransposeMap = decltype(make_layout(
+      ShapeV{}, make_stride(get<1>(StrideV{}), get<0>(StrideV{}), get<2>(StrideV{}))));
+  using SLayoutV = decltype(composition(SLayoutK{}, TransposeMap{}));
+
+  // using SwizzleOp = Swizzle<3, 4, 3>;
```
Comment (shaochangxu) [inline src/attention/prefill/config.h]: 不用的代码删除吧

File: src/attention/prefill/config.h [hunk 3]
```diff
@@ -276,6 +459,112 @@ struct AttentionKVCachePrefillFp8Config {
   auto get_shm_size() { return shm_size; }
 };
 
+template <typename Tin_, typename Tout_, typename TiledMmaQKAtom, typename TiledMmaPVAtom,
+          int kTileM_, int kTileN_, int kTileK_, int kTileV_, int kTileS_, int kBlockSize_,
+          int kScaleBlockSize_, int kStage_, int kWarpgroupM_ = 2, int kWarpgroupN_ = 1,
+          int kSwizzleQ = 128, int kSwizzleK = 128, int kSwizzleV = 128, int kSwizzleY = 128>
+struct AttentionKVCachePrefillQKTokenVTenorFp8Config {
```
Comment (shaochangxu) [inline src/attention/prefill/config.h]: 这几个config有大量代码是一样的，是不是抽出来一个公共的，然后整几个继承公共的就行

File: src/attention/prefill/kernels.cuh [hunk 4]
```diff
@@ -241,6 +205,79 @@ __global__ void update_batched_tma_with_kvcache(const vec_t<cute::TmaDescriptor,
   }
 }
 
+// TMA update kernel for FP8 varlen Q/K/V.
+template <typename Tin, typename Tout, typename TmaQ, typename TmaK, typename TmaV, typename TmaY>
+__global__ void update_batched_tma_fp8(const vec_t<cute::TmaDescriptor, 4> td_qkvy,
```
Comment (shaochangxu) [inline src/attention/prefill/kernels.cuh]: 这个和之前的update tma区别是啥？fp8有啥特殊的吗

File: src/attention/prefill/kernels.cuh [hunk 5]
```diff
@@ -1561,94 +1638,2083 @@ __global__ void attention_prefill_bf16_multi_stage_kernel(cute::TmaDescriptor *t
...
+        cute::copy(tma_y.with(td_y), tYss(_, iwarpgroup, 0),
+                   tYgg(_, itile_m * 2 + iwarpgroup, 0, ihead_q));
+        tma_store_arrive();
+      }
+    }
+  }
+}
+
+template <typename Config, typename TmaQ, typename TmaK, typename TmaV, typename TmaY>
+__global__ void attention_prefill_bf16_multi_stage_kernel(cute::TmaDescriptor *td_qkvy,
```
Comment (shaochangxu) [inline src/attention/prefill/kernels.cuh:3124]: 这个kernel是和blocksparse有关的吗？之前没有的话也不用新加？

File: tests/test_attention_blocksparse_prefill_fp8_dim192.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+        cu_seqlens_kv,
+        softmax_scale,
+        q_descale=q_scale.item(),
+        k_descale=k_scale.item(),
+        v_descale=v_scale.item(),
+        causal=True,
+        block_mask=block_mask,
+    )
+
+    out = hpc.attention_blocksparse_prefill_fp8_dim192(
```
Comment (shaochangxu) [inline tests/test_attention_blocksparse_prefill_fp8_dim192.py]: test 也是不需要给dim192一个新文件，原来的test case里加就行

File: tests/test_attention_blocksparse_qpertoken_perhead_kvpertensor_fp8.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+        num_seq,
+        quant_type=hpc.QuantType.QPERTOKEN_PERHEAD_KPERTENSOR_VPERTENSOR,
+        block_mask=block_mask_u8,
+    )
+
+    assert allclose(gt, my, atol=0.1)
+
+
+@pytest.mark.parametrize("skip_ratio", [0.0, 0.5])
+def test_blocksparse_qpertoken_perhead_kvpertensor_fp8_fixed_pscale(skip_ratio):
```
Comment (shaochangxu) [inline tests/test_attention_blocksparse_qpertoken_perhead_kvpertensor_fp8.py]: fixed pscale不需要一个单独的单测

File: tests/test_fuse_moe_pertensor.py [hunk 8]
```diff
@@ -83,14 +83,10 @@ def naive_group_gemm(x, w, cu_seqlens, scale, expert_ids):
...
 
-        y_group = torch._scaled_mm(
-            x_group,
-            w_group.t(),
-            scale_a=scale[i],
-            scale_b=torch.ones((1), dtype=torch.float, device="cuda"),
-            bias=None,
-            out_dtype=torch.bfloat16,
-        )
+        y_group = (
```
Comment (shaochangxu) [inline tests/test_fuse_moe_pertensor.py:86]: 为啥还改moe的内容

File: tests/test_attention_with_kvcache_qpertoken_perhead_kvpertensor_prefill_fp8.py [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+    return output
+
+
+@pytest.mark.parametrize("kv_layout", ["nhd"])
+@pytest.mark.parametrize("num_batch", [4])
+@pytest.mark.parametrize("num_seq_q", [128, 256, 512])
+@pytest.mark.parametrize("num_seq_kv", [1024])
+@pytest.mark.parametrize("num_head_q", [4, 16])
+@pytest.mark.parametrize("num_head_kv", [1, 4])
+def test_attention_with_kvcache_prefill_fp8_fixed_pscale(
```
Comment (shaochangxu) [inline tests/test_attention_with_kvcache_qpertoken_perhead_kvpertensor_prefill_fp8.py]: 同上，pscale不用加额外的单测，另外内部的仓库也同步改一波

File: tests/test_fuse_moe_pertensor.py (hunk 8 above)
Comment (bcacdwk) [reply tests/test_fuse_moe_pertensor.py:86]: torch._scaled_mm 在torch210 cuda130的时候，用了 cuBLAS fp8 GEMM，要求 mat2 的 N 维必须是 16 的倍数。MoE 测试的 token 数可能是 6、9 等非 16 倍数，会触发 CUBLAS_STATUS_NOT_SUPPORTED
所以这里用来验证 kernel 正确性的 ground truth 没必要走 fp8 cuBLAS 路径，换成 fp32 matmul 就不报错了

File: hpc/attention.py (hunk 1 above)
Comment (bcacdwk) [reply hpc/attention.py]: 这里最开始是dispatch的方式，但是主要在于这个是面向DSV31的，不仅仅是dim变成192，里面的KV变成了varlen的格式，quantization也是全per-tensor的，和dim128的路径区别比较大，所以之前内部的版本迭代又重新把他们拆开了

File: src/attention/prefill/kernels.cuh (hunk 4 above)
Comment (bcacdwk) [reply src/attention/prefill/kernels.cuh]: 因为之前的update_batched_tma是必须要求Sq=Skv的，这个是支持chunked版本。我重新提交一个，删掉这个新的，把这个功能追加到前面的update tma

File: tests/test_attention_blocksparse_prefill_fp8_dim192.py (hunk 6 above)
Comment (bcacdwk) [reply tests/test_attention_blocksparse_prefill_fp8_dim192.py]: 这个主要还是考虑量化方法、head dim都不一样，因为之前不同的两种量化方法也是拆开写test case的，所以这里也拆开了。因为这个相当于是额外单独的量化（all per-tensor）

File: tests/test_attention_blocksparse_qpertoken_perhead_kvpertensor_fp8.py (hunk 7 above)
Comment (bcacdwk) [reply tests/test_attention_blocksparse_qpertoken_perhead_kvpertensor_fp8.py]: 好的 随后删除

File: src/attention/prefill/config.h (hunk 2 above)
Comment (bcacdwk) [reply src/attention/prefill/config.h]: 好的，这里内部仓库也同步简化

File: src/attention/prefill/config.h (hunk 3 above)
Comment (bcacdwk) [reply src/attention/prefill/config.h]: 把所有的slayout_selector全部提取出来了，删掉了大量的重复代码
typo 也会修复 (Tenor -> Tensor)

File: src/attention/prefill/kernels.cuh (hunk 5 above)
Comment (bcacdwk) [reply src/attention/prefill/kernels.cuh:3124]: 这里我检查了一下，应该是cuh前面插入了较多的内容，在这里diff被判定为都是新的内容，但是实际上这里都没改

File: tests/test_attention_with_kvcache_qpertoken_perhead_kvpertensor_prefill_fp8.py (hunk 9 above)
Comment (bcacdwk) [reply tests/test_attention_with_kvcache_qpertoken_perhead_kvpertensor_prefill_fp8.py]: 好的

**Review by bcacdwk — COMMENTED** [review]
reply commit
